File: java/pattern/schema/Rule.java
/*
 * Copyright (C) 2021 Vaticle
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.vaticle.typeql.lang.pattern.schema;

import com.vaticle.typeql.lang.common.exception.TypeQLException;
import com.vaticle.typeql.lang.pattern.Conjunction;
import com.vaticle.typeql.lang.pattern.Definable;
import com.vaticle.typeql.lang.pattern.Disjunction;
import com.vaticle.typeql.lang.pattern.Negation;
import com.vaticle.typeql.lang.pattern.Pattern;
import com.vaticle.typeql.lang.pattern.variable.Reference;
import com.vaticle.typeql.lang.pattern.variable.ThingVariable;
import com.vaticle.typeql.lang.pattern.variable.Variable;

import javax.annotation.Nullable;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.vaticle.typeql.lang.common.TypeQLToken.Char.COLON;
import static com.vaticle.typeql.lang.common.TypeQLToken.Char.CURLY_CLOSE;
import static com.vaticle.typeql.lang.common.TypeQLToken.Char.CURLY_OPEN;
import static com.vaticle.typeql.lang.common.TypeQLToken.Char.SEMICOLON;
import static com.vaticle.typeql.lang.common.TypeQLToken.Char.SPACE;
import static com.vaticle.typeql.lang.common.TypeQLToken.Schema.RULE;
import static com.vaticle.typeql.lang.common.TypeQLToken.Schema.THEN;
import static com.vaticle.typeql.lang.common.TypeQLToken.Schema.WHEN;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_THEN;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_THEN_HAS;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_THEN_VARIABLES;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_WHEN_CONTAINS_DISJUNCTION;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_WHEN_MISSING_PATTERNS;
import static com.vaticle.typeql.lang.common.exception.ErrorMessage.INVALID_RULE_WHEN_NESTED_NEGATION;

public class Rule implements Definable {
    private final String label;
    private Conjunction<? extends Pattern> when;
    private ThingVariable<?> then;
    private int hash = 0;

    public Rule(String label) {
        this.label = label;
    }

    public Rule(String label, Conjunction<? extends Pattern> when, ThingVariable<?> variable) {
        validate(label, when, variable);
        this.label = label;
        this.when = when;
        this.then = variable;
    }

    @Override
    public boolean isRule() {
        return true;
    }

    @Override
    public Rule asRule() {
        return this;
    }

    public String label() {
        return label;
    }

    public Conjunction<? extends Pattern> when() {
        return when;
    }

    public ThingVariable<?> then() {
        return then;
    }

    public IncompleteRule when(Conjunction<? extends Pattern> when) {
        return new IncompleteRule(label, when);
    }

    public static void validate(String label, Conjunction<? extends Pattern> when, ThingVariable<?> then) {
        validateWhen(label, when);
        validateThen(label, when, then);
    }

    private static void validateWhen(String label, Conjunction<? extends Pattern> when) {
        if (when == null) throw new NullPointerException("Null when pattern");
        if (when.patterns().size() == 0) throw TypeQLException.of(INVALID_RULE_WHEN_MISSING_PATTERNS.message(label));
        if (findNegations(when).anyMatch(negation -> findNegations(negation.pattern()).findAny().isPresent())) {
            throw TypeQLException.of(INVALID_RULE_WHEN_NESTED_NEGATION.message(label));
        }
        if (findDisjunctions(when).findAny().isPresent()) {
            throw TypeQLException.of(INVALID_RULE_WHEN_CONTAINS_DISJUNCTION.message(label));
        }
    }

    private static Stream<Negation<?>> findNegations(Pattern pattern) {
        if (pattern.isNegation()) return Stream.of(pattern.asNegation());
        if (pattern.isVariable()) return Stream.empty();
        return pattern.patterns().stream().flatMap(Rule::findNegations);
    }

    private static Stream<Disjunction<?>> findDisjunctions(Pattern pattern) {
        if (pattern.isDisjunction()) return Stream.of(pattern.asDisjunction());
        if (pattern.isVariable()) return Stream.empty();
        return pattern.patterns().stream().flatMap(Rule::findDisjunctions);
    }

    private static void validateThen(String label, @Nullable Conjunction<? extends Pattern> when, ThingVariable<?> then) {
        if (then == null) throw new NullPointerException("Null then pattern");
        int numConstraints = then.constraints().size();

        // rules must contain contain either 1 has constraint, or a isa and relation constrain
        if (!((numConstraints == 1 && then.has().size() == 1) || (numConstraints == 2 && then.relation().isPresent() && then.isa().isPresent()))) {
            throw TypeQLException.of(INVALID_RULE_THEN.message(label, then));
        }

        // rule 'has' cannot assign both an attribute type and a named variable
        if (then.has().size() == 1 && then.has().get(0).type().isPresent() && then.has().get(0).attribute().isNamed()) {
            throw TypeQLException.of(INVALID_RULE_THEN_HAS.message(label, then));
        }

        // all user-written variables in the 'then' must be present in the 'when', if it exists.
        if (when != null) {
            Set<Reference> thenReferences = Stream.concat(Stream.of(then), then.variables())
                    .filter(Variable::isNamed).map(Variable::reference).collect(Collectors.toSet());

            Set<Reference> whenReferences = when.variables()
                    .filter(Variable::isNamed).map(Variable::reference).collect(Collectors.toSet());

            if (!whenReferences.containsAll(thenReferences)) {
                throw TypeQLException.of(INVALID_RULE_THEN_VARIABLES.message(label));
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder syntax = new StringBuilder();
        syntax.append(RULE).append(SPACE).append(label);

        if (when != null || then != null) syntax.append(COLON).append(SPACE);

        // when
        if (when != null) {
            syntax.append(WHEN).append(SPACE).append(CURLY_OPEN).append(SPACE);
            for (Pattern p : when.patterns()) {
                syntax.append(p).append(SEMICOLON).append(SPACE);
            }
            syntax.append(CURLY_CLOSE).append(SPACE);
        }

        // then
        if (then != null) {
            syntax.append(THEN).append(SPACE).append(CURLY_OPEN).append(SPACE);
            syntax.append(then).append(SEMICOLON).append(SPACE);
            syntax.append(CURLY_CLOSE);
        }

        return syntax.toString();
    }

    @Override
    public int hashCode() {
        if (hash == 0) {
            this.hash = Objects.hash(label, when, then);
        }
        return hash;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rule that = (Rule) o;
        return (this.label.equals(that.label) &&
                Objects.equals(this.when, that.when) &&
                Objects.equals(this.then, that.then));
    }

    public static class IncompleteRule {
        private final String label;
        private final Conjunction<? extends Pattern> when;

        public IncompleteRule(String label, Conjunction<? extends Pattern> when) {
            this.label = label;
            this.when = when;
        }

        public Rule then(ThingVariable<?> then) {
            return new Rule(label, when, then);
        }
    }
}

File: rust/pattern/schema/rule.rs
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use std::{collections::HashSet, fmt, iter};

use crate::{
    common::{
        error::{collect_err, TypeQLError},
        token,
        validatable::Validatable,
        Result,
    },
    pattern::{Conjunction, HasConstraint, Pattern, ThingStatement, VariablesRetrieved},
    variable::variable::VariableRef,
    Label,
};

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct RuleLabel {
    pub label: Label,
}

impl RuleLabel {
    pub fn new(label: Label) -> Self {
        RuleLabel { label }
    }

    pub fn when(self, when: Conjunction) -> RuleLabelWhen {
        RuleLabelWhen { label: self.label, when }
    }
}

impl Validatable for RuleLabel {
    fn validate(&self) -> Result {
        Ok(())
    }
}

impl<T: Into<Label>> From<T> for RuleLabel {
    fn from(label: T) -> Self {
        RuleLabel::new(label.into())
    }
}

impl fmt::Display for RuleLabel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {}", token::Schema::Rule, self.label)
    }
}

pub struct RuleLabelWhen {
    pub label: Label,
    pub when: Conjunction,
}

impl RuleLabelWhen {
    pub fn then(self, then: ThingStatement) -> Rule {
        Rule { label: self.label, when: self.when, then }
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Rule {
    pub label: Label,
    pub when: Conjunction,
    pub then: ThingStatement,
}

impl Validatable for Rule {
    fn validate(&self) -> Result {
        collect_err([
            validate_no_nested_negations(self.when.patterns.iter(), &self.label),
            validate_valid_inference(&self.then, &self.label),
            validate_then_bounded_by_when(&self.then, &self.when, &self.label),
            self.when.validate(),
            self.then.validate(),
        ])
    }
}

fn validate_no_nested_negations<'a>(patterns: impl Iterator<Item = &'a Pattern>, rule_label: &Label) -> Result {
    collect_err(patterns.map(|p| -> Result {
        match p {
            Pattern::Conjunction(c) => validate_no_nested_negations(c.patterns.iter(), rule_label),
            Pattern::Statement(_) => Ok(()),
            Pattern::Disjunction(d) => validate_no_nested_negations(d.patterns.iter(), rule_label),
            Pattern::Negation(n) => {
                if contains_negations(iter::once(n.pattern.as_ref())) {
                    Err(TypeQLError::InvalidRuleWhenNestedNegation { rule_label: rule_label.clone() })?
                } else {
                    Ok(())
                }
            }
        }
    }))
}

fn contains_negations<'a>(mut patterns: impl Iterator<Item = &'a Pattern>) -> bool {
    patterns.any(|p| match p {
        Pattern::Conjunction(c) => contains_negations(c.patterns.iter()),
        Pattern::Statement(_) => false,
        Pattern::Disjunction(d) => contains_negations(d.patterns.iter()),
        Pattern::Negation(_) => true,
    })
}

fn validate_valid_inference(then: &ThingStatement, rule_label: &Label) -> Result {
    if infers_ownership(then) {
        let has = then.has.get(0).unwrap();
        if let HasConstraint::HasConcept(Some(type_label), attr_var) = has {
            Err(TypeQLError::InvalidRuleThenHas {
                rule_label: rule_label.clone(),
                then: then.clone(),
                variable: attr_var.clone(),
                type_label: type_label.clone(),
            })?
        }
        Ok(())
    } else if infers_relation(then) {
        let relation = then.relation.as_ref().unwrap();
        if !relation.role_players.iter().all(|rp| rp.role_type.is_some()) {
            Err(TypeQLError::InvalidRuleThenRoles { rule_label: rule_label.clone(), then: then.clone() })?
        }
        Ok(())
    } else {
        Err(TypeQLError::InvalidRuleThen { rule_label: rule_label.clone(), then: then.clone() })?
    }
}

fn infers_ownership(then: &ThingStatement) -> bool {
    then.has.len() == 1
        && (then.iid.is_none() && then.isa.is_none() && then.predicate.is_none() && then.relation.is_none())
}

fn infers_relation(then: &ThingStatement) -> bool {
    then.relation.is_some()
        && then.isa.is_some()
        && (then.iid.is_none() && then.has.is_empty() && then.predicate.is_none())
}

fn validate_then_bounded_by_when(then: &ThingStatement, when: &Conjunction, rule_label: &Label) -> Result {
    let bounds: HashSet<VariableRef<'_>> = when.retrieved_variables().collect();
    if !then.variables().filter(|r| r.is_name()).all(|r| bounds.contains(&r)) {
        Err(TypeQLError::InvalidRuleThenVariables { rule_label: rule_label.clone() })?
    }
    Ok(())
}

impl fmt::Display for Rule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} {}: {} {} {} {{\n    {};\n}}",
            token::Schema::Rule,
            self.label,
            token::Schema::When,
            self.when,
            token::Schema::Then,
            self.then
        )
    }
}
